### DISCUSSION TITLE
musb universal interface

### DISCUSSION THREAD

**Discussion by ReimuNotMoe** [hathach/tinyusb#1267] [Ideas]
The Mentor musb IP core is used by many MCUs & MPUs. But currently the implementation in TinyUSB is hardcoded to only support some TI MCUs. Is there any way to make it support more processors? (Just like the glue layers implemented by the Linux `musb.c`)

**Comment by hathach** [comment]
yeah, I did something similar for synopsys dwc2 here
- https://github.com/hathach/tinyusb/tree/master/src/portable/synopsys/dwc2 

The reason why it is specific to TI is because we haven't got any other mcu with same IP to do the generalization. Luckily, there is an on-going allwinner f1c200 PR #1220 , which also use musb IP. Once PR is merged, I will try to do the generalization (since not everyone have all required boards to generalize and verify it work on all ports). 

which MCU you are trying to port TinyUSB to ? You could duplicate the current musb driver, then replace all the hw,registers with your mcu. I could help to generalize it later on.

**Reply by ReimuNotMoe** [reply]
I'm going to port it to the PIC32MZ(EC/EF/DA) platform.

For the PIC devices:
- PIC32MZEC, PIC32MZEF, PIC32MZDA use the high speed dual role musb IP, 8x2 EPs max
- PIC32MX, PIC32MM, PIC32MK, PIC32MZW1, PIC24, dsPIC33 use the same full speed dual role USB IP, 16x2 EPs max
- PIC18 devices use a full speed device mode only USB IP, 16x2 EPs max (maybe a cut down version of the above)
- PIC16 devices use a full speed device mode only USB IP, 8x2 EPs max (maybe a cut down version of the above)

Other devices using the musb IP:
- TI AM35x & OMAP2430, and some Davinci DSPs
- Ingenic JZ4740
- ST-Ericsson Ux500

Thank you for the information and happy new year!

**Comment by hathach** [comment]
thank you for your info, it sounds great. I think @kasjer is also working on PIC32MZ as well. for PIC32MM, it uses the same IP as nxp kinetis family as well as mindmotion mm32. I have a plan to generalize it as chipidea fullspeed in the future. If you did a port for PIC32MM, please check it out 
- https://github.com/hathach/tinyusb/pull/1086#issuecomment-924910217
- https://github.com/hathach/tinyusb/blob/master/src/portable/nxp/khci/dcd_khci.c

Happy New Year to you as well :tada:

**Reply by kasjer** [reply]
@ReimuNotMoe, @hathach, I do have working PIC32MZEF high speed device role code. I will present PR soon. Host and OTG was not touched.

**Reply by hathach** [reply]
superb! great works on this, I am looking forward to see your PR :100:

**Reply by ReimuNotMoe** [reply]
So there's another problem to discuss: When it comes to PIC32 software development, there are 2 fashions: (1) Using all stock Microchip tools and compilers, such as the MPLAB XC32 (a GNU toolchain with some private modifications + closed source libc) (2) Use the unmodified GNU toolchain + free software libc (e.g. newlib). And I'm using the second one.

However it's impossible to make one codebase compatible with both fashions without extensive uses of macros. Here's an interrupt handling example:

MPLAB XC32:
```
void __attribute__((vector(_USB_VECTOR), interrupt(ipl7srs), nomips16)) USB_ISR() {
    // ...
}
```

GCC:
```
void __attribute__((interrupt, use_shadow_register_set, __section__(".vectors"))) USB_ISR() {
    // ...
}
```

In my personal opinion, since MPLAB XC32 is not open source friendly (you actually need to purchase its license in order to use optimization levels other than -O1 & -O2), we should support the second way by default. 

Do you have any idea on this?
